rust/cymbal/src/symbol_store.rs

```diff
@@ -0,0 +1,193 @@
+use std::{sync::Arc, time::Duration};
+
+use axum::async_trait;
+use reqwest::Url;
+use tracing::{info, warn};
+
+use crate::{config::Config, error::Error};
+
+#[async_trait]
+pub trait SymbolStore: Send + Sync + 'static {
+    // Symbol stores return an Arc, to allow them to cache (and evict) without any consent from callers
+    async fn fetch(&self, team_id: i32, r: SymbolSetRef) -> Result<Arc<Vec<u8>>, Error>;
+}
+
+pub enum SymbolSetRef {
+    Js(Url),
+}
+
+// A store that implements basic lookups, for whatever that means for each language. In
+// JS, that means it does fetching and searches for sourcemap references. In other languages,
+// it might mean talking to S3, or something else. It implements no caching, and no storing of
+// fetched symbol sets - other stores should wrap this one to provide that functionality.
+pub struct Basic {
+    pub client: reqwest::Client,
+}
+
+impl Basic {
+    pub fn new(config: &Config) -> Result<Self, Error> {
+        let mut client = reqwest::Client::builder();
+
+        let timeout = Duration::from_secs(config.sourcemap_timeout_seconds);
+
+        if !config.allow_internal_ips {
+            client = client.dns_resolver(Arc::new(common_dns::PublicIPv4Resolver {}));
+        } else {
+            warn!("Internal IPs are allowed, this is a security risk");
+        }
+
+        let client = client.timeout(timeout).build()?;
+
+        Ok(Self { client })
+    }
+}
+
+#[async_trait]
+impl SymbolStore for Basic {
+    async fn fetch(&self, _: i32, r: SymbolSetRef) -> Result<Arc<Vec<u8>>, Error> {
+        let SymbolSetRef::Js(sref) = r; // We only support this
+        let Some(sourcemap_url) = find_sourcemap_url(&self.client, sref.clone()).await? else {
+            // TODO - this might not actually count as an error, and simply means we don't /need/ a sourcemap
+            // for a give frame, but I haven't decided how to handle that yet
+            return Err(Error::InvalidSourceRef(format!(
+                "No sourcemap URL found for {}",
+                sref
+            )));
+        };
+        fetch_source_map(&self.client, sourcemap_url)
+            .await
+            .map(Arc::new)
+    }
+}
+
+async fn find_sourcemap_url(client: &reqwest::Client, start: Url) -> Result<Option<Url>, Error> {
+    info!("Fetching sourcemap from {}", start);
+    let res = client.get(start).send().await?;
+    // First, we check for the sourcemap headers: SourceMap, or X-SourceMap
+    let headers = res.headers();
+    let header_url = headers
+        .get("SourceMap")
+        .or_else(|| headers.get("X-SourceMap"));
+
+    if let Some(header_url) = header_url {
+        info!("Found sourcemap header: {:?}", header_url);
+        let url = header_url.to_str().map_err(|_| {
+            Error::InvalidSourceRef(format!("Failed to parse url from header of {}", res.url()))
+        })?;
+        let url: Url = url
+            .parse()
```

Comment:
I wonder if this value will be relative like it can be in the `sourceMappingURL`? I would expect that this parsing will fail in that case

Reply:
Gah, it can be, of course 🤦 

Why is the web like this, will fix

Reply:
Fixed